### ISSUE TITLE
Nested jar URLs cannot be split and reassembled resulting in errors with projects that use this technique (such as JobRunr)

### ISSUE THREAD

**Issue by rdehuyss** [spring-projects/spring-boot#38592]
Hi, I'm Ronald, the creator of [JobRunr](https://github.com/jobrunr/jobrunr).

[An issue](https://github.com/jobrunr/jobrunr/issues/884) was reported to us regarding the new nested jar support which broke JobRunr as it reads files from the classpath.

Inside an Uber jar, it is currently impossible to list the files within a directory using a [NIO FileSystemProvider](https://docs.oracle.com/javase/8/docs/api/java/nio/file/spi/FileSystemProvider.html)

I would expect the following to work:

```java
var url = ((JarURLConnection)H2StorageProvider.class.getResource("migrations").openConnection()).getJarFileURL();
FileSystem fileSystem = FileSystems.newFileSystem(url.toURI(), Collections.emptyMap());
Path path = fileSystem.getPath("BOOT-INF/lib/jobrunr-6.3.3.jar!/org/jobrunr/storage/sql/common/migrations/");
boolean exists = Files.exists(path);
List<Path> list = Files.list(path).collect(Collectors.toList());
return "Root exists: " + exists + "; contents: " + list;
```

The FileSystem returned is of type class org.springframework.boot.loader.nio.file.NestedFileSystem.

For a reproducible scenario, please see https://github.com/rdehuyss/spring-boot-3.2-nested-jar-issue

**Comment by rdehuyss** [comment]
Tagging @philwebb as this related to:
- https://github.com/spring-projects/spring-boot/issues/37668
- https://github.com/spring-projects/spring-boot/commit/7ad4a9817da4b04c1a743cc52a7585dd30b25ef9

**Comment by philwebb** [comment]
Thanks very much for the detailed report @rdehuyss. I think we've actually got two distinct bugs here so I've opened #38595 for the second.

I think I have a fix for the [sample project](https://github.com/oleh-melnyk/example-java-mag/tree/issue/884) provided in https://github.com/jobrunr/jobrunr/issues/884 but I'm not sure about the [reproducer](https://github.com/rdehuyss/spring-boot-3.2-nested-jar-issue) you attached to this issue.

As it stands, that your sample is calling `H2StorageProvider.class.getResource("migrations")` which returns a URL in the form `jar:nested:/the.jar/!BOOT-INF/lib/jobrunr-6.3.3.jar!/org/jobrunr/storage/sql/h2/migrations`.

It then gets the JAR file URL which is `nested:/the.jar/!BOOT-INF/lib/jobrunr-6.3.3.jar` and creates a `FileSystem` from it. Finally, it calls `getPath("BOOT-INF/lib/jobrunr-6.3.3.jar!/org/jobrunr/storage/sql/common/migrations/")`.

There are a couple of problems here:
1) The filesystem created is not a `ZipFileSystem`
2) The path include the nested jar in it.

If I change the sample as follows, things work.

```java
@GetMapping("/get/from/nested/jar")
public String getFromNestedJar() throws IOException, URISyntaxException {
	loadFileSystemIfNecessary();
	Path path = fileSystem.getPath("org/jobrunr/storage/sql/common/migrations/");
	boolean exists = Files.exists(path);
	List<Path> list = Files.list(path).collect(Collectors.toList());
	return "Root exists: " + exists + "; contents: " + list;
}

private static void loadFileSystemIfNecessary() throws IOException, URISyntaxException {
	if (fileSystem == null) {
		URL resource = H2StorageProvider.class.getResource("migrations");
		var jarFileUrl = ((JarURLConnection) resource.openConnection()).getJarFileURL();
		fileSystem = FileSystems.newFileSystem(new URI("jar:" + jarFileUrl), Collections.emptyMap());
	}
}
```

I think this makes sense because if we remove nested JARs from the picture, the original sample wouldn't work. For example, the following code:

```java
URL url = new URL("jar:file:/the.jar!/");
JarURLConnection connection = (JarURLConnection) url.openConnection();
URL jarFileUrl = connection.getJarFileURL();
FileSystem fileSystem = FileSystems.newFileSystem(jarFileUrl.toURI(), Collections.emptyMap());
```

Fails when `FileSystems.newFileSystem` is called with:

```
Exception in thread "main" java.lang.IllegalArgumentException: Path component should be '/'
	at java.base/sun.nio.fs.UnixFileSystemProvider.checkUri(UnixFileSystemProvider.java:81)
	at java.base/sun.nio.fs.UnixFileSystemProvider.newFileSystem(UnixFileSystemProvider.java:90)
	at java.base/java.nio.file.FileSystems.newFileSystem(FileSystems.java:339)
	at java.base/java.nio.file.FileSystems.newFileSystem(FileSystems.java:288)
	at org.spring.boot.nested.Test.main(Test.java:16)
```

In other words, we always need the `jar:` scheme if we want to look at the jar contents.

I hope that makes sense and the fix works OK. If it doesn't, please let me know.

**Comment by rdehuyss** [comment]
Hi @philwebb - thanks for your swift intervention and input.

Just to be 100% sure, I changed the reproducer project to the above example you gave (still using Spring Boot 3.2.0). Is it possible I then encounter the issue about [java.util.zip.ZipException: read CEN tables failed](https://github.com/spring-projects/spring-boot/issues/38595)? I assume this will then be fixed in Spring Boot 3.2.1?

**Comment by philwebb** [comment]
That correct. There's a 3.2.1-SNAPSHOT available if you want to give that a try. It's in repo.spring.io/snapshot. You can use start.spring.io to generate a project with the maven config you need.

**Comment by wilkinsona** [comment]
@rdehuyss The latest 3.2.1-SNAPSHOT, available from https://repo.spring.io/snapshot, should contain both of Phil's recent changes. Could you please give it a try?

**Comment by rdehuyss** [comment]
Will do! Any idea on a release date already?

**Comment by bclozel** [comment]
@rdehuyss it's scheduled on December 21st, see https://github.com/spring-projects/spring-boot/milestones and https://calendar.spring.io

**Comment by rdehuyss** [comment]
I should have known that there is a release calendar 🤓.

**Comment by rdehuyss** [comment]
Hi @philwebb , @wilkinsona, @bclozel  :  I can confirm that JobRunr works again with Spring Boot 3.2.1-SNAPSHOT. 

Has anybody already mentioned you guys are awesome?! Thanks for the swift handling and feedback. I ❤️ that I do not need to change anything in JobRunr 🎉 (which I was expecting).
